src/devices/Mcp3008/Mcp3008.cs

```diff
@@ -29,41 +23,36 @@ public void Dispose()
             _spiDevice = null;
         }
 
-        public int Read(int channel, InputConfiguration inputConfiguration = InputConfiguration.SingleEnded)
+        public int Read(int channel)
         {
             if (channel < 0 || channel > 7)
             {
                 throw new ArgumentException("ADC channel must be within 0-7 range.");
             }
 
-            return ReadSpi(channel, inputConfiguration);
-        }
+            Span<byte> writeBuffer = stackalloc byte[3];
+            Span<byte> readBuffer = stackalloc byte[3];
 
-        private static byte GetConfigurationBits(int channel, InputConfiguration inputConfiguration)
-        {
-            int configurationBits = (0b0001_1000 | channel) << 3;
+            // first we send 5 bits, rest is ignored
+            // start bit = 1
+            // conversion bit = 1 (single ended)
+            // next 3 bits are channel
+            writeBuffer[0] = (byte)((0b1_1000 | channel) << 3);
+            _spiDevice.TransferFullDuplex(writeBuffer, readBuffer);
 
-            if (inputConfiguration == InputConfiguration.Differential)
+            // we ignore first 7 bits then read following 10
+            // 7 bits:
+            // - 5 are used for writing start bit, conversion bit and channel
+            // - next bit is ignored so device has time to sample
+            // - following bit should be 0
+            if ((readBuffer[0] & 0b10) != 0)
             {
-                configurationBits &= 0b1011_1111; // Clear mode bit.
+                throw new InvalidOperationException("Invalid data was read from the sensor");
             }
 
-            return (byte)configurationBits;
-        }
-
-        // Ported: https://github.com/adafruit/Adafruit_Python_MCP3008/blob/master/Adafruit_MCP3008/MCP3008.py
-        private int ReadSpi(int channel, InputConfiguration inputConfiguration)
-        {
-            byte configurationBits = GetConfigurationBits(channel, inputConfiguration);
-            byte[] writeBuffer = new byte[] { configurationBits, 0, 0 };
-            byte[] readBuffer = new byte[3];
-
-            _spiDevice.TransferFullDuplex(writeBuffer, readBuffer);
-
-            int result = (readBuffer[0] & 0b0000_0001) << 9;
-            result |= (readBuffer[1] & 0b1111_1111) << 1;
-            result |= (readBuffer[2] & 0b1000_0000) >> 7;
-            result = result & 0b0011_1111_1111;
+            int result = (readBuffer[0] & 1) << 9;
+            result |= readBuffer[1] << 1;
+            result |= readBuffer[2] >> 7;
```

Comment:
This is different than what we were doing before. Before we were doing `(readBuffer[2] & 0b1000_0000) >> 7;` but you are flipping all of readBuffer[2] bits. Also, you seem to have missed the clearing of a couple of bits that we used to do here: `result = result & 0b0011_1111_1111;`

Reply:
yes, I got read of clearing bits because they effectively didn't do anything.

I.e. masking byte with `0b1111_1111` is the same as not doing anything since it's already fulfilling the mask.
Shifting by 7 to the right - you will lose first 7 LSB bits anyway so there is no point in extra bitmasking

Reply:
that part makes sense, but you did remove the other part `result = result & 0b0011_1111_1111;` which was clearing a couple of bits in the result, that part I believe we should still keep.

Reply:
@joperezr the way we currently shift they will always be zero anyway

Reply:
I see you are right.